test/hack/bin/verify-rules.sh

```diff
@@ -0,0 +1,166 @@
+        git ls-files |
+            sed -En 's_^helm/prometheus-rules/(templates/alerting-rules/.*\.ya?ml)$_\1_p' || echo error
+    )
+
+    # Get prefixes whitelisted via the failure_file
+    local -a expected_failure_prefixes_global=()
+    [[ -f "$expected_failure_file_global" ]] \
+        && mapfile -t expected_failure_prefixes_global <"$expected_failure_file_global"
+
+    local -a providers
+    mapfile -t providers <"$GIT_WORKDIR/test/conf/providers"
+
+    local -a promtool_check_errors=()
+    local -a promtool_test_errors=()
+
+    for provider in "${providers[@]}"; do
+        echo "### Running tests for provider: $provider"
+
+        # Get the list of whitelisted files for this provider
+        local expected_failure_relative_file_provider="test/conf/promtool_ignore_$provider"
+        local expected_failure_file_provider="$GIT_WORKDIR/$expected_failure_relative_file_provider"
+        local -a expected_failure_prefixes_provider=()
+        [[ -f "$expected_failure_file_provider" ]] \
+            && mapfile -t expected_failure_prefixes_provider <"$expected_failure_file_provider"
+
+        for file in "${all_files[@]}"; do
+
+            # if the file is whitelisted via the global ignore file
+            if array_contains "$file" "${expected_failure_prefixes_global[@]}"; then
+                # don't run tests for it
+                echo "###  Skipping $file: listed in $expected_failure_relative_file_global"
+                continue
+            fi
+            # if the file is whitelisted via the provider ignore file
+            if array_contains "$file" "${expected_failure_prefixes_provider[@]}"; then
+                # don't run tests for it
+                echo "###  Skipping $file: listed in $expected_failure_relative_file_provider"
+                continue
+            fi
+
+            echo "###  Testing $file"
+
+            # retrieve basename in pure bash
+            filename="${file##*/}"
+
+            # don't run tests if no provider specific tests are defined
+            if [[ ! -d "$GIT_WORKDIR/test/tests/providers/$provider" ]]; then
+                echo "###   No tests for proviter $provider - skipping"
+                continue
+            fi
+
+            echo "###    extracting $GIT_WORKDIR/test/providers/$provider/$filename"
+            "$GIT_WORKDIR/$HELM" template \
+                --set="managementCluster.provider.kind=$provider" \
+                --release-name prometheus-rules \
+                --namespace giantswarm "$GIT_WORKDIR"/helm/prometheus-rules \
+                -s "$file" |
+                "$GIT_WORKDIR/$YQ" '.spec' - >"$GIT_WORKDIR/test/tests/providers/$provider/$filename"
+
+            echo "###    promtool check rules $GIT_WORKDIR/test/tests/providers/$provider/$filename"
+            local promtool_check_output
+            promtool_check_output="$("$GIT_WORKDIR/$PROMTOOL" check rules "$GIT_WORKDIR/test/tests/providers/$provider/$filename" 2>&1)" ||
```

Comment:
to be more strict, we could think about doing the `promtool check rules` command (validates the syntax of the rules) for every rules file and doing the unit-testing later on.

But maybe this is hopefully a motivation for everyone getting rule file validation by implementing unit tests :sweat_smile:

Reply:
I thought about that, but that means having another set of ignore files, because some rule files can't even be extracted from helm, or depend on the cloudprovider.

On the other hand, it may not be useful to do the `prom check rules` for each cloudprovider. Maybe we could just do:
- create a list to be checked
- first provider, check all we can and clean them from the list, but if helm extract fails we skip them and keep them in the list to be tried with the next provider.

This process would not care about ignore lists.

Yes, I'll try to do it!